starcheck/state_checks.py

```diff
@@ -0,0 +1,184 @@
+import numpy as np
+from cxotime import CxoTime
+
+from astropy.table import Table, vstack
+import astropy.units as u
+from functools import lru_cache
+import numpy as np
+
+import kadi.commands as kadi_commands
+import kadi.commands.states as kadi_states
+from cxotime import CxoTime
+from chandra_maneuver import duration
+from Quaternion import Quat
+
+
+@lru_cache
+def make_man_table():
+    """
+    Calculate the time for range of manuever angles using Chandra.Maneuver
+    duration.
+    """
+    durations = []
+    q0 = Quat(equatorial=(0, 0, 0))
+
+    # Use a range of angles that sample the curve pretty well by eye
+    # The duration function is a little slow and not vectorized, so
+    # this is sparse.
+    angles = [0, 5, 10, 15, 20, 25, 35, 50, 100, 150, 180]
+    for angle in angles:
+        q1 = Quat(equatorial=(angle, 0, 0))
+        durations.append(duration(q0, q1))
+
+    return Table([durations, angles], names=["duration", "angle"])
+
+
+def calc_pcad_state_nman_sums(states):
+    """
+    For each NPNT state in a monotonically increasing list of states,
+    calculate the sum of time in NMAN before it
+    and interpolate into the man_table to calculate the equivalent
+    angle for a duration.
+    """
+    man_table = make_man_table()
+
+    nman_sums = []
+    single_nman_sum = 0
+    for state in states:
+        if state["pcad_mode"] == "NMAN":
+            single_nman_sum += state["tstop"] - state["tstart"]
+        if (state["pcad_mode"] == "NPNT") & (single_nman_sum > 0):
+            nman_sums.append(
+                {
+                    "tstart": state["tstart"],
+                    "nman_sum": single_nman_sum.copy(),
+                    "angle": np.interp(
+                        single_nman_sum, man_table["duration"], man_table["angle"]
+                    ),
+                }
+            )
+            # Reset the sum of NMM time to 0 on any NPNT state
+            single_nman_sum = 0
+
+        # If in a funny state, just add a large "maneuver time" of 20ks
+        if state["pcad_mode"] not in ["NMAN", "NPNT"]:
+            single_nman_sum += 20000
```

Comment:
That's generally fair. I figured there is no guarantee about what mode we're in if extra states are needed for continuity before the start of loads (not actually the weekly loads) and it was probably better for this code to do *something* with that case but in practice shouldn't matter.

Reply:
Operationally any maneuver of interest must be preceded by a NPNT (the recovery attitude). So any intervals in NSUN can safely be ignored in all situations. In effect the safing recovery process is guaranteed to zero out the maneuver errror so ignoring NSUN is OK. 

If that wasn't the case then that magic 20000 number would not be adequate to account for potentially very large (degrees) maneuver errors.

Reply:
I think it might make sense to just add up the time in not NPNT before each NPNT state so weird ones wouldn't need anything special in continuity.  But that probably wouldn't work as well with your idea about replacing the logic to add this up with just using remerged merged states.